addons/addon-base-raas/packages/base-raas-services/lib/aws-accounts/aws-cfn-service.js

```diff
@@ -305,6 +339,96 @@ class AwsCfnService extends Service {
+      { action: 'attempt-onboard', conditions: [allowIfActive, allowIfAdmin] },
+      {},
+    );
+
+    const awsAccountsService = await this.service('awsAccountsService');
+    const accounts = await awsAccountsService.list();
+    const pendingAccountIds = _.map(
+      _.filter(accounts, acct => acct.permissionStatus === 'PENDING'),
+      'id',
+    );
+
+    const auditLog = {};
+
+    const processor = async awsAccountId => {
+      try {
+        await this.finishOnboardingAccount(requestContext, awsAccountId);
+        auditLog[awsAccountId] = 'Successfully Onboarded';
+      } catch (e) {
+        auditLog[awsAccountId] = `Account is not ready yet. ${e}`;
+      }
+    };
+
+    // For each account, reach out 10 at a time
+    if (!_.isEmpty(pendingAccountIds)) {
+      await processInBatches(pendingAccountIds, 10, processor);
+    } else {
+      auditLog.status = 'No accounts are pending.';
+    }
+
+    await this.audit(requestContext, {
+      action: 'onboard-aws-accounts',
+      body: auditLog,
+    });
+    return auditLog;
+  }
+
+  async finishOnboardingAccount(requestContext, accountId) {
+    await this.assertAuthorized(
+      requestContext,
+      { action: 'check-aws-permissions', conditions: [allowIfActive, allowIfAdmin] },
+      { accountId },
+    );
+    const PENDING_STATES = [
+      'CREATE_IN_PROGRESS',
+      'UPDATE_IN_PROGRESS',
+      'UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS',
+      'UPDATE_ROLLBACK_IN_PROGRESS',
+      'REVIEW_IN_PROGRESS',
+    ];
+    const awsAccountsService = await this.service('awsAccountsService');
+    const accountEntity = await awsAccountsService.mustFind(requestContext, { id: accountId });
+
+    const region = this.settings.get(settingKeys.awsRegion);
+    const { onboardStatusRoleArn, cfnStackName, externalId } = accountEntity;
+    const cfnApi = await this.getCfnSdk(onboardStatusRoleArn, externalId, region);
+    const params = { StackName: cfnStackName };
+    const stacks = await cfnApi.describeStacks(params).promise();
+    const stack = _.find(_.get(stacks, 'Stacks', []), item => item.StackName === cfnStackName);
+
+    if (_.isEmpty(stack)) {
+      throw this.boom.notFound(`Stack '${cfnStackName}' not found`, true);
+    }
+
+    if (PENDING_STATES.includes(stack.StackStatus)) {
+      throw this.boom.badRequest(`Stack '${cfnStackName}' is still pending.`, true);
+    }
+
+    const fieldsToUpdate = {};
+    const findOutputValue = prop => {
+      const output = _.find(_.get(stack, 'Outputs', []), item => item.OutputKey === prop);
+      return output.OutputValue;
+    };
+
+    fieldsToUpdate.cfnStackId = stack.StackId;
+    fieldsToUpdate.externalId = accountEntity.externalId || 'workbench';
+    fieldsToUpdate.vpcId = findOutputValue('VPC');
+    fieldsToUpdate.subnetId = findOutputValue('VpcPublicSubnet1');
+    fieldsToUpdate.encryptionKeyArn = findOutputValue('EncryptionKeyArn');
+    fieldsToUpdate.roleArn = findOutputValue('CrossAccountEnvMgmtRoleArn');
```

Comment:
There should be 3 roles in total in member accounts including: onboard status role, envMgmtRole and crossAccountRole. Are they saved in DDB after onboarding?

Reply:
only two, onboardStatusRole and the roleArn one. onboardStatusRole is saved in the account's entry in the DDB when it is first created, and then roleArn is stored in this method (it's pulling the CrossAccountEnvMgmtRoleArn value). The third arn shouldn't be necessary to store at the account level.

Reply:
xAccEnvMgmtRole is needed, it's used for Service Catalog related calls in member account, the only case it's not needed is if the member account is the same as the main account. You can do a [global search](https://github.com/awslabs/service-workbench-on-aws/search?q=xAccEnvMgmtRoleArn) and see its use case.

Reply:
nvm, resolved with the other comment